actionpack/lib/action_dispatch/middleware/cookies.rb

```diff
@@ -633,14 +668,26 @@ def initialize(parent_jar)
       private
         def parse(name, encrypted_message)
           deserialize(name) do |rotate|
-            @encryptor.decrypt_and_verify(encrypted_message, on_rotation: rotate)
+            if cookies_with_metadata?
+              value = @encryptor.decrypt_and_verify(encrypted_message, on_rotation: rotate, purpose: name)
+              if value.nil? && read_cookies_without_metadata?
+                value = @encryptor.decrypt_and_verify(encrypted_message, on_rotation: rotate)
+              end
+              value
+            else
+              @encryptor.decrypt_and_verify(encrypted_message, on_rotation: rotate)
+            end
           end
         rescue ActiveSupport::MessageEncryptor::InvalidMessage, ActiveSupport::MessageVerifier::InvalidSignature
           parse_legacy_signed_message(name, encrypted_message)
         end
 
         def commit(options)
-          options[:value] = @encryptor.encrypt_and_sign(serialize(options[:value]), expiry_options(options))
+          if cookies_with_metadata?
+            options[:value] = @encryptor.encrypt_and_sign(serialize(options[:value]), fetch_metadata(options))
```

Comment:
Why is this method concerned with `cookies_with_metadata?`? Especially when there's a method that's called `fetch_metadata`? Same for the other instances.